module/zfs/dbuf.c

```diff
@@ -2285,7 +2317,7 @@ dbuf_dirty(dmu_buf_impl_t *db, dmu_tx_t *tx)
 
 	dprintf_dbuf(db, "size=%llx\n", (u_longlong_t)db->db.db_size);
 
-	if (db->db_blkid != DMU_BONUS_BLKID) {
+	if (db->db_blkid != DMU_BONUS_BLKID && db->db_state != DB_NOFILL) {
 		dmu_objset_willuse_space(os, db->db.db_size, tx);
```

Comment:
I am not very intimate with this code, but I have suspicion the same should be added for `dr->dr_accounted = db->db.db_size;` assignment below, otherwise dbuf_undirty() and may be other places may break the math.